### DISCUSSION TITLE
How to redirect from an async page function?

### DISCUSSION THREAD

**Discussion by espeoceka** [zauberzeug/nicegui#549] [Q&A] [answered]
Hey im not even sure its issue, i might be doing something wrong....

For example:

```py
@ui.page('/logout')
async def logout(request: Request) -> None:
    if await is_authenticated(request):
        await redis.expire(request.session["id"], 0)
        request.session['id'] = None
        return RedirectResponse('/login')
    else:
        return RedirectResponse('/')
```

It never redirect to anything, but when i remove async it works like intended. Any ideas how to sort this out ?

**Comment by falkoschindler** [comment, marked as answer]
Thanks for reaching out, @espeoceka!

I was confused at first, since it is working without `async`. But now I'm actually not sure, why it is working at all...
As the type annotation says, `logout` is supposed to build a UI and return nothing. But instead it returns a FastAPI response. So `logout` isn't a NiceGUI page function, but a regular FastAPI endpoint. Therefore it should be decorated with `app.get` instead of `ui.page`.

For example, the following code is working with and without `async`:

```py
@app.get('/redirect')
async def redirect() -> None:
    return RedirectResponse('/')
```

**Reply by ed2050** [reply]
It's not stated that RedirectResponse is a _fastapi_ class.  Code above doesn't work unless you add this:
```
from fastapi.responses import RedirectResponse
```
see https://fastapi.tiangolo.com/uk/advanced/custom-response/#redirectresponse

**Comment by falkoschindler** [comment]
@rodja and I just noticed that @espeoceka's code example actually _should_ work with the current implementation of `ui.page`. Therefore I opened an issue #693 to fix it.

**Comment by espeoceka** [comment]
Hey great that this has been worked on, i found also issue related to redirecting. This looks atrocious on first glance but im trying to make dependency that simply authorize user and if no user profile is found redirects to login......

```py
@ui.page('/user/contact_lists')
async def user_contact_lists(request: Request,client:Client,db_session=Depends(get_db),user_profile=Depends(get_user_profile)):
    while True:
        status = await client.connected(timeout=5)
        if not client.is_waiting_for_connection:
            break
        await asyncio.sleep(1)
    
    if not isinstance(user_profile,User):
        ui.open(f"/login?destination={client.page.path}")
```
This "charming" loop somehow makes redirection works.... but it cannot be used in dependency due to some pydantic error. I already guessing its not good approach at best. How this should be done ?

PS. I tried to make a dependency that takes Client class but it seems to be impossible due to pydantic error ( Client class added to arguments of get_user_profile )

**Reply by rodja** [reply]
Could you provide a minimal reproducible example? That would help us to test and understand your problem.

**Reply by espeoceka** [reply]
This should give same results:

```py
@ui.page('/user/contact_lists')
async def user_contact_lists(request: Request,client:Client,user_profile=Depends(get_user_profile)):
    while True:
        status = await client.connected(timeout=5)
        if not client.is_waiting_for_connection:
            break
        await asyncio.sleep(1)
    
    if not user_profile:
        ui.open(f"/login?destination={client.page.path}")

async def get_user_profile(request:Request,client:Client):
    return False
```
 If client is removed from get_user_profile then no errors but then without that loop any redirection is not working

**Reply by falkoschindler** [reply]
Thanks! It looks like the problem is already apparent in the following simplified example:

```py
from fastapi import Depends
from nicegui import Client, ui

def get_client_id(client: Client) -> str:
    return client.id

@ui.page('/')
def user_contact_lists(get_client_id=Depends(get_client_id)):
    print(get_client_id)

ui.run()
```

> fastapi.exceptions.FastAPIError: Invalid args for response field! Hint: check that <class 'nicegui.client.Client'> is a valid Pydantic field type. If you are using a return type annotation that is not a valid Pydantic field (e.g. Union[Response, dict, None]) you can disable generating the response model from the type annotation with the path operation decorator parameter response_model=None. Read more: https://fastapi.tiangolo.com/tutorial/response-model/

FastAPI tries to resolve the `get_client_id` dependency, which in turn requires a `Client`. But since `Client` is no Pydantic class, it cannot be resolved.

**Reply by espeoceka** [reply]
Yes, exacly and another thing is if we decide to get rid off Client from dependency that loop is required else redirections dont work ( i suppose this happens only with async )

**Reply by rodja** [reply]
I think we need to figure out how to make the client created in 

https://github.com/zauberzeug/nicegui/blob/3d1ee70fa6c9dde43a1fc8f4720e201180a9ca0a/nicegui/page.py#L66

available to FastAPIs

https://github.com/tiangolo/fastapi/blob/c81e136d75f5ac4252df740b35551cf2afb4c7f1/fastapi/dependencies/utils.py#L546

**Reply by falkoschindler** [reply]
Annoyingly FastAPI seems to resolve dependencies before entering our page function. Therefore `global.get_client()` fails with an IndexError if we try to do it like this:
```py
def get_client_id() -> str:
    return globals.get_client().id

@ui.page('/')
def user_contact_lists(client_id=Depends(get_client_id, use_cache=False)):
    print(client_id)
```

**Reply by falkoschindler** [reply]
We're still thinking about a way to support FastAPI's dependency injection with our `Client` object. But in principle it's tricky, because the dependency injection runs before entering the page function, but only the page function creates the client. And if other functions depend on the client (like `get_user_profile`), it's not clear how to resolve it, because the client depends on the current request.

**Reply by espeoceka** [reply]
Yea i thought this might be this sad one....

Maybe its possible to declare dependecies on ui.page level instead ? ( they can be solved there and their results can be passed down to function as regular kwargs )

**Reply by ysharoiko** [reply]
No progress on that?
I have also stucked
```py
#!/usr/bin/env python3.11
#-*- coding: utf-8 -*-

from nicegui import ui, background_tasks
from fastapi.responses import RedirectResponse
import asyncio

@ui.page('/download')
def index_page():
  async def load():
    await asyncio.sleep(3) # fetch data or something else which takes time (must be async)
    
    content.clear() 
    with content:
      ui.markdown('Done. Redirecting ...')

    #ui.add_head_html('''
    #<script>
    #  window.location.href = window.location.protocol + "//" + window.location.host
    #</script>
#  ''')
    ui.open(f"http://google.ru")
    #return RedirectResponse("http://google.ru")
      
  with ui.element("div").style("align-self: center; padding-top:100pt;"):
    with ui.card().style(f"background-color: #dedede, align-items: center;") as content:
      ui.spinner('dots', size='5em', color='#81b983') 
      #size='10em'
      ui.label("Preparing...").style("padding-top:18pt;")

      background_tasks.create(load())

ui.run()
```

None of above redirecting is working :(
